<h1><center>Assignment 4

Name: Sharimbaev Batir

**You need to work with data from from the about 128 years of Olympic history by pandas. Write code and perform computations in the cells below. There are 5 questions.** 

You will use datasets which used in previous assignment and you can download the files `winter.csv` and `summer.csv` from [here](https://drive.google.com/drive/folders/1k6BtheuRBaLcsF7698-kolc3LZFocaIN?usp=sharing).

In [2]:
import pandas as pd

In [3]:
PATH1 = 'summer.csv'
PATH2 = 'winter.csv'

In [4]:
dataSummer = pd.read_csv(PATH1)
dataSummer

FileNotFoundError: [Errno 2] No such file or directory: 'summer.csv'

In [5]:
dataWinter = pd.read_csv(PATH2)
dataWinter

FileNotFoundError: [Errno 2] No such file or directory: 'winter.csv'

<b>1. Merge dataframes: dataSummer and dataWinter with all join types (inner, outer, left, right) and write what was the difference

In [36]:
pd.merge(dataWinter, dataSummer, how='outer')

Year      City      Sport           Discipline  \
0      1924  Chamonix   Biathlon             Biathlon   
1      1924  Chamonix   Biathlon             Biathlon   
2      1924  Chamonix   Biathlon             Biathlon   
3      1924  Chamonix   Biathlon             Biathlon   
4      1924  Chamonix   Biathlon             Biathlon   
...     ...       ...        ...                  ...   
36930  2012    London  Wrestling  Wrestling Freestyle   
36931  2012    London  Wrestling  Wrestling Freestyle   
36932  2012    London  Wrestling  Wrestling Freestyle   
36933  2012    London  Wrestling  Wrestling Freestyle   
36934  2012    London  Wrestling  Wrestling Freestyle   

                         Athlete Country Gender            Event   Medal  
0                    BERTHET, G.     FRA    Men  Military Patrol  Bronze  
1                 MANDRILLON, C.     FRA    Men  Military Patrol  Bronze  
2            MANDRILLON, Maurice     FRA    Men  Military Patrol  Bronze  
3                VANDELLE, André     FRA    Men  Military Patrol  Bronze  
4           AUFDENBLATTEN, Adolf     SUI    Men  Military Patrol    Gold  
...                          ...     ...    ...              ...     ...  
36930         JANIKOWSKI, Damian     POL    Men         Wg 84 KG  Bronze  
36931  REZAEI, Ghasem Gholamreza     IRI    Men         Wg 96 KG    Gold  
36932             TOTROV, Rustam     RUS    Men         Wg 96 KG  Silver  
36933          ALEKSANYAN, Artur     ARM    Men         Wg 96 KG  Bronze  
36934             LIDBERG, Jimmy     SWE    Men         Wg 96 KG  Bronze  

[36935 rows x 9 columns]

In [37]:
pd.merge(dataWinter, dataSummer, how='inner')

Empty DataFrame
Columns: [Year, City, Sport, Discipline, Athlete, Country, Gender, Event, Medal]
Index: []

In [1]:
pd.merge(dataWinter, dataSummer, how='left')

NameError: name 'pd' is not defined

In [39]:
pd.merge(dataWinter, dataSummer, how='right')

Year    City      Sport           Discipline  \
0      1896  Athens   Aquatics             Swimming   
1      1896  Athens   Aquatics             Swimming   
2      1896  Athens   Aquatics             Swimming   
3      1896  Athens   Aquatics             Swimming   
4      1896  Athens   Aquatics             Swimming   
...     ...     ...        ...                  ...   
31160  2012  London  Wrestling  Wrestling Freestyle   
31161  2012  London  Wrestling  Wrestling Freestyle   
31162  2012  London  Wrestling  Wrestling Freestyle   
31163  2012  London  Wrestling  Wrestling Freestyle   
31164  2012  London  Wrestling  Wrestling Freestyle   

                         Athlete Country Gender                       Event  \
0                  HAJOS, Alfred     HUN    Men              100M Freestyle   
1               HERSCHMANN, Otto     AUT    Men              100M Freestyle   
2              DRIVAS, Dimitrios     GRE    Men  100M Freestyle For Sailors   
3             MALOKINIS, Ioannis     GRE    Men  100M Freestyle For Sailors   
4             CHASAPIS, Spiridon     GRE    Men  100M Freestyle For Sailors   
...                          ...     ...    ...                         ...   
31160         JANIKOWSKI, Damian     POL    Men                    Wg 84 KG   
31161  REZAEI, Ghasem Gholamreza     IRI    Men                    Wg 96 KG   
31162             TOTROV, Rustam     RUS    Men                    Wg 96 KG   
31163          ALEKSANYAN, Artur     ARM    Men                    Wg 96 KG   
31164             LIDBERG, Jimmy     SWE    Men                    Wg 96 KG   

        Medal  
0        Gold  
1      Silver  
2      Bronze  
3        Gold  
4      Silver  
...       ...  
31160  Bronze  
31161    Gold  
31162  Silver  
31163  Bronze  
31164  Bronze  

[31165 rows x 9 columns]

Outer: Use union of keys from both frames. So we have 36935 rows, because dataWinter has 5770 rows and dataSummer has
31165 rows. Therefore 5770 + 31165 = 36935

Inner: Use intersection of keys from both frames. The elements in A are not in B and the elements in B are not in A.
So we have not dataset.

Left: We know dW intersection dS equal to zero. Therefore We have only use keys of left frame. It is dataWinter.

Right: We know dW intersection dS equal to zero. Therefore We have only use keys of right frame. It is dataSummer.

<b>2. Create a hierarchical index on the concatenation of two dataframes and count missing values

In [40]:
df2 = pd.concat([dataWinter, dataSummer ], keys=['Winter','Summer'],axis=1)
df2.head()

Winter                                                                      \
     Year      City     Sport Discipline               Athlete Country Gender   
0  1924.0  Chamonix  Biathlon   Biathlon           BERTHET, G.     FRA    Men   
1  1924.0  Chamonix  Biathlon   Biathlon        MANDRILLON, C.     FRA    Men   
2  1924.0  Chamonix  Biathlon   Biathlon   MANDRILLON, Maurice     FRA    Men   
3  1924.0  Chamonix  Biathlon   Biathlon       VANDELLE, André     FRA    Men   
4  1924.0  Chamonix  Biathlon   Biathlon  AUFDENBLATTEN, Adolf     SUI    Men   

                           Summer                               \
             Event   Medal   Year    City     Sport Discipline   
0  Military Patrol  Bronze   1896  Athens  Aquatics   Swimming   
1  Military Patrol  Bronze   1896  Athens  Aquatics   Swimming   
2  Military Patrol  Bronze   1896  Athens  Aquatics   Swimming   
3  Military Patrol  Bronze   1896  Athens  Aquatics   Swimming   
4  Military Patrol    Gold   1896  Athens  Aquatics   Swimming   

                                                                          
              Athlete Country Gender                       Event   Medal  
0       HAJOS, Alfred     HUN    Men              100M Freestyle    Gold  
1    HERSCHMANN, Otto     AUT    Men              100M Freestyle  Silver  
2   DRIVAS, Dimitrios     GRE    Men  100M Freestyle For Sailors  Bronze  
3  MALOKINIS, Ioannis     GRE    Men  100M Freestyle For Sailors    Gold  
4  CHASAPIS, Spiridon     GRE    Men  100M Freestyle For Sailors  Silver

In [41]:
sum(df2.isnull().values.ravel())

228559

<b>3.1. Create dataframe which contains information about year, city, country and medal from both datasets (dataSummer and dataWinter) and delete duplicated information from dataframe

In [42]:
df3 = pd.merge(dataWinter, dataSummer, how='outer')
df3

Year      City      Sport           Discipline  \
0      1924  Chamonix   Biathlon             Biathlon   
1      1924  Chamonix   Biathlon             Biathlon   
2      1924  Chamonix   Biathlon             Biathlon   
3      1924  Chamonix   Biathlon             Biathlon   
4      1924  Chamonix   Biathlon             Biathlon   
...     ...       ...        ...                  ...   
36930  2012    London  Wrestling  Wrestling Freestyle   
36931  2012    London  Wrestling  Wrestling Freestyle   
36932  2012    London  Wrestling  Wrestling Freestyle   
36933  2012    London  Wrestling  Wrestling Freestyle   
36934  2012    London  Wrestling  Wrestling Freestyle   

                         Athlete Country Gender            Event   Medal  
0                    BERTHET, G.     FRA    Men  Military Patrol  Bronze  
1                 MANDRILLON, C.     FRA    Men  Military Patrol  Bronze  
2            MANDRILLON, Maurice     FRA    Men  Military Patrol  Bronze  
3                VANDELLE, André     FRA    Men  Military Patrol  Bronze  
4           AUFDENBLATTEN, Adolf     SUI    Men  Military Patrol    Gold  
...                          ...     ...    ...              ...     ...  
36930         JANIKOWSKI, Damian     POL    Men         Wg 84 KG  Bronze  
36931  REZAEI, Ghasem Gholamreza     IRI    Men         Wg 96 KG    Gold  
36932             TOTROV, Rustam     RUS    Men         Wg 96 KG  Silver  
36933          ALEKSANYAN, Artur     ARM    Men         Wg 96 KG  Bronze  
36934             LIDBERG, Jimmy     SWE    Men         Wg 96 KG  Bronze  

[36935 rows x 9 columns]

In [43]:
df3.drop(['Sport', 'Discipline', 'Athlete', 'Gender', 'Event'], axis=1, inplace=True)
df3

Year      City Country   Medal
0      1924  Chamonix     FRA  Bronze
1      1924  Chamonix     FRA  Bronze
2      1924  Chamonix     FRA  Bronze
3      1924  Chamonix     FRA  Bronze
4      1924  Chamonix     SUI    Gold
...     ...       ...     ...     ...
36930  2012    London     POL  Bronze
36931  2012    London     IRI    Gold
36932  2012    London     RUS  Silver
36933  2012    London     ARM  Bronze
36934  2012    London     SWE  Bronze

[36935 rows x 4 columns]

In [44]:
df3.drop_duplicates(inplace=True)
df3

Year      City Country   Medal
0      1924  Chamonix     FRA  Bronze
4      1924  Chamonix     SUI    Gold
8      1924  Chamonix     FIN  Silver
12     1924  Chamonix     BEL  Bronze
21     1924  Chamonix     GBR  Silver
...     ...       ...     ...     ...
36902  2012    London     GEO  Bronze
36908  2012    London     EST  Silver
36910  2012    London     TUR  Bronze
36924  2012    London     ARM  Silver
36933  2012    London     ARM  Bronze

[3413 rows x 4 columns]

<b> 3.2. Add to created dataframe full country name (Ex: Country: FRA, FullCountry: France) (for first 10)

In [45]:
df3.head(10)

Year      City Country   Medal
0   1924  Chamonix     FRA  Bronze
4   1924  Chamonix     SUI    Gold
8   1924  Chamonix     FIN  Silver
12  1924  Chamonix     BEL  Bronze
21  1924  Chamonix     GBR  Silver
31  1924  Chamonix     GBR    Gold
39  1924  Chamonix     SWE  Silver
47  1924  Chamonix     GBR  Bronze
59  1924  Chamonix     CAN    Gold
68  1924  Chamonix     USA  Silver

In [46]:
counter1 = 0
index = []
for i in df3.index:
    if counter1 < 10:
        index.append(i)
    counter1 += 1

counter2 = 0
country = []
for i in df3['Country']:
    if counter2 < 10:
        country.append(i)
    counter2 += 1
    
for i in range(len(index)):
    if country[i] == 'FRA':
        df3.at[index[i], 'Country'] = 'France'
    elif country[i] == 'SUI':
        df3.at[index[i], 'Country'] = 'Switzerland'
    elif country[i] == 'FIN':
        df3.at[index[i], 'Country'] = 'Finland'
    elif country[i] == 'BEL':
        df3.at[index[i], 'Country'] = 'Belgium'
    elif country[i] == 'GBR':
        df3.at[index[i], 'Country'] = 'Great Britain'
    elif country[i] == 'SWE':
        df3.at[index[i], 'Country'] = 'Sweden'
    elif country[i] == 'CAN':
        df3.at[index[i], 'Country'] = 'Canada'
    elif country[i] == 'USA':
        df3.at[index[i], 'Country'] = 'United States of America' 
    
df3.head(10)

Year      City                   Country   Medal
0   1924  Chamonix                    France  Bronze
4   1924  Chamonix               Switzerland    Gold
8   1924  Chamonix                   Finland  Silver
12  1924  Chamonix                   Belgium  Bronze
21  1924  Chamonix             Great Britain  Silver
31  1924  Chamonix             Great Britain    Gold
39  1924  Chamonix                    Sweden  Silver
47  1924  Chamonix             Great Britain  Bronze
59  1924  Chamonix                    Canada    Gold
68  1924  Chamonix  United States of America  Silver

<b>4. Create dataframe with random values (more than 1000 data). You can specify shape as you want and filter data from outliers

In [1]:
import numpy as np

In [48]:
np.random.seed(190103206)
df4 = pd.DataFrame(np.random.randn(2500, 4))
df4.describe()

0            1            2            3
count  2500.000000  2500.000000  2500.000000  2500.000000
mean      0.000209     0.029780    -0.009786    -0.005335
std       1.006907     1.011515     1.034168     1.005567
min      -3.643103    -3.165688    -3.730673    -3.688408
25%      -0.685290    -0.683192    -0.706686    -0.700555
50%      -0.020493     0.034057    -0.034105    -0.016184
75%       0.682462     0.744709     0.676553     0.690415
max       3.721111     3.260412     4.081165     3.394902